x-pack/legacy/plugins/maps/public/embeddable/map_embeddable.tsx

```diff
@@ -83,6 +83,7 @@ export interface MapEmbeddableInput extends EmbeddableInput {
   mapCenter?: MapCenter;
   hiddenLayers?: string[];
   hideFilterActions?: boolean;
+  layerList?: unknown[];
```

Comment:
I don't really like adding `layerList` to `MapEmbeddableInput` type. From my understanding, `EmbeddableInput` is a collection of embeddable state that can be modified while in the embeddable container and is stored separate to that of the saved object. For example, the map embeddable allows for storing a center location that is independent of the map saved object center location. This allows the same map saved object to be embedded in many dashboards and each dashboard can independently store the center so they can each have a different starting location. `layerList` does not really fit this description of mutable state stored by the embeddable container. It fells its getting put into `MapEmbeddableInput` because it has no better place to go.

Reply:
>  From my understanding, EmbeddableInput is a collection of embeddable state that can be modified while in the embeddable container

yes

> and is stored separate to that of the saved object.

Not necessarily.  Input represents the serialized state of the embeddable. You should be able to recreate the embeddable by persisting the input. Keeping layerList separate from input will cause this heuristic to fail in the case for SIEM when they are trying to use map embeddable without a saved object backing it.

I've been working on coming up with better examples of embeddables by value and embeddables by reference.  I have a PR here: https://github.com/elastic/kibana/pull/61749/.

I add an example where an embeddable can optionally be both, something that fits with the "save to library" flow visualizations are going with.  The  input for my example `TodoSoEmbeddable` is:

{ 
  savedObjectId?: string
  attributes: SavedObjectAttributes
}

The output is
{ 
  savedAttributes: SavedObjectAttributes
}

The input attributes can override the saved object attributes, the user can edit and save to the saved object, or just save "by value" with the container.

If you don't like going this route, another option is to expose another embeddable type where the input does take { layerList }, and that can be the "by value" embeddable that SIEM uses.  Or you could also expose a function on your plugin contract just for SIEM, something like `createMapEmbeddable(layerList, input)`. 

Happy to chat about these concepts tomorrow.

Reply:
i think embeddables should be created from non saved state. creating them from saved state should be done by utility function (embeddable factory in our case) and saving that state should be done by corresponding saved object type. That makes object id an optional property on the embeddable, which is only there for easy consumption of saved object type, but should rather not be used directly inside embeddable. i am wondering if it should even be there, or is the caller of factory method responsible for memorizing that if it needs it for later ? (editor or dashboard)

but could we still provide additional state, which is not supposed to change, to the constructor ? for example the layerList above, or if we decide to keep the savedObjectId ?

but

Reply:
Chatted with Peter offline.  Some reasons to move savedObjectId as input: 

 - Container no longer has special knowledge about whether an embeddable is backed by a saved object or not. That is up to the implementation of each embeddable.
- Generic actions can act on any embeddable that has `SavedObjectEmbeddableInput`  (e.g. has `savedObjectId` and `savedObjectType`).  Right now savedObjectId is only part of the container panels, not inside input. It is not consistently represented in a uniform manner with an embeddable.  This PR (https://github.com/elastic/kibana/pull/61692/files) creates the `SavedObjectEmbeddableInput`  type.  Now you could create simple generic actions for any embeddable that is backed by a saved object - no matter whether they are in a container or just by themselves. 

 We'll continue to discuss within the App Arch team about savedObjectId. For the meantime I moved layerList outside of input and made it a setter.